README.md
# hubot-vk

----

**IMPORTANT: LOOKING FOR A DEVELOPER, create [an issue](https://gitlab.com/volebo/hubot-vk/issues)**

**[CONTRIBUTIONS](https://gitlab.com/volebo/hubot-vk/pulls) ARE WELCOME**

----

This is a [Hubot](http://hubot.github.com/) adapter to use with [vk.com](https://vk.com).

[![Travis branch](https://img.shields.io/travis/volebo/hubot-vk/master.svg?maxAge=2592000)](https://travis-ci.org/volebo/hubot-vk)
[![npm](https://img.shields.io/npm/v/hubot-vk.svg)][npm]
[![npm downloads](https://img.shields.io/npm/dm/hubot-vk.svg)][npm]
[![codecov](https://codecov.io/gh/volebo/hubot-vk/branch/master/graph/badge.svg)](https://codecov.io/gh/volebo/hubot-vk)
[![codebeat badge](https://codebeat.co/badges/96144a51-6304-4f09-b687-5aa313223497)](https://codebeat.co/projects/gitlab-com-volebo-hubot-vk)

Currently, we have just maintainers, but no developers...

Comprehensive documentation [is NOT available](https://volebo.gitlab.io/hubot-vk).

## License

Please, read the [`LICENSE`](LICENSE) file in the root of the repository (or downloaded package).

[npm]: https://www.npmjs.com/package/hubot-vk


CHANGELOG.md
# CHANGELOG

```
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
```

--------------------------------------------------------------------------------
>
> All notable changes to this project will be documented in this file.
>
> The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
> and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
>
> **This file MUST be filled only by maintainers, using messages from pull**
> **requests.**
>
> new release template
>
> ## [Unreleased] - yyyy-mm-dd
>
> ### Added
> ### Changed
> ### Fixed
> ### Removed
>
--------------------------------------------------------------------------------

## [0.0.6] - 2017-01-31

### Removed

- no more GulpJS

## [0.0.5] - 2017-01-31

:camel:

## [0.0.4] - 2017-01-31

:sweat_smile: nothing changed, but I've updated dependencies and repo-structure.

## [0.0.3] - 2017-01-31

:sweat_smile: nothing changed.

## [0.0.2] - 2017-01-27

:sweat_smile: nothing changed.

## [0.0.1] - 2017-01-27

INIT.


tests/bootstrap.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# A VK.COM adapter for hubot.
#
# Copyright (C) 2016-2020 Volebo <dev@volebo.net>
# Copyright (C) 2016-2020 Maksim Koryukov <maxkoryukov@volebo.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

'use strict'

const chai                         = require('chai')
const path                         = require('path')
const mh                           = require('@volebo/mocha-helpers')


global.expect = chai.expect

global.filename2suitename = mh.filename2suitename
global.tags = mh.tags


const report_path = path.join(
	process.cwd(),
	'tmp',
	'mochawesome-report'
)

process.env.MOCHAWESOME_REPORTDIR = report_path


src/package.spec.js
'use strict'

const pkg = require('../')

describe('package.spec', () => {
	it('should export a function', () => {
		expect(pkg).to.be.a('function')
	})
})


src/package.test.js
'use strict'

const pkg = require('../')

describe('package.test', () => {
	it('always pass', () => {
		expect(true).is.true
		expect(pkg).exist
	})
})


.mocharc.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# A VK.COM adapter for hubot.
#
# Copyright (C) 2016-2020 Volebo <dev@volebo.net>
# Copyright (C) 2016-2020 Maksim Koryukov <maxkoryukov@volebo.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

'use strict'

module.exports = {
	reporter: 'list',
	// reporter: 'spec',

	recursive: true,
	checkLeaks: true,
	ui: 'bdd',
	require: 'tests/bootstrap',
	extension: ['test.js'],

	diff: true,
	//opts: './test/mocha.opts',
	package: './package.json',
	slow: 75,
	timeout: 2000,

	spec: [
		'src/**/*.test.js',
	]
}


nyc.config.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# A VK.COM adapter for hubot.
#
# Copyright (C) 2016-2020 Volebo <dev@volebo.net>
# Copyright (C) 2016-2020 Maksim Koryukov <maxkoryukov@volebo.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

'use strict'

module.exports = {
	'all': true,

	// "sourceMap": false,
	// "instrument": false,

	'reporter': [
		'html',
		'lcov',
		'text',
		'text-summary',
		// 'json-summary',
	],

	'report-dir': './tmp/coverage',

	'include': [
		'src/**',
	],

	'exclude': [
		'**/*.test.js',
		'**/*.spec.js',
		'tests/bootstrap.js',
	],
	// ["coverage/**","packages/*/test/**","test/**","test{,-*}.js","**/*{.,-}test.js","**/__tests__/**","**/{ava,babel,jest,nyc,rollup,webpack}.config.js"]]

	'excludeNodeModules': true
}


src/adapter.js
"use strict"

const HTTPS         = require("https");

const hubot         = require("hubot")

const Robot = hubot.robot
const Adapter = hubot.adapter
const Message = hubot.message

const TextMessage   =  Message.TextMessage
const EnterMessage  =  Message.EnterMessage
const LeaveMessage  =  Message.LeaveMessage
const TopicMessage  =  Message.TopicMessage

const Streaming = require("./streaming")

const utils = require("./utils")
const slice = utils.slice

class VkAdapter extends Adapter {

	constructor(robot) {
		super(robot)
        @verifyToken = process.env.MESSENGER_VERIFY_TOKEN
        @accessToken = process.env.MESSENGER_ACCESS_TOKEN
        @apiURL = "https://graph.facebook.com/v2.6"
        @robot.logger.info "hubot-messenger-bot: Adapter loaded."
	}

	send {
		var envelope, string, strings;
		envelope = arguments[0], strings = 2 <= arguments.length ? slice.call(arguments, 1) : [];
		if (strings.length > 0) {
			string = strings.shift();
			if (typeof string === 'function') {
				string();
				return this.send.apply(this, [envelope].concat(slice.call(strings)));
			} else {
				return this.bot.Room(envelope.room).speak(string, (function(_this) {
					return function(err, data) {
						if (err != null) {
							_this.robot.logger.error("Campfire send error: " + err);
						}
						return _this.send.apply(_this, [envelope].concat(slice.call(strings)));
					};
				})(this));
			}
		}
	}

	emote() {
		var envelope, strings;
		envelope = arguments[0], strings = 2 <= arguments.length ? slice.call(arguments, 1) : [];
		return this.send.apply(this, [envelope].concat(slice.call(strings.map(function(str) {
			return "*" + str + "*";
		}))));
	}

	reply() {
		var envelope, strings;
		envelope = arguments[0], strings = 2 <= arguments.length ? slice.call(arguments, 1) : [];
		return this.send.apply(this, [envelope].concat(slice.call(strings.map(function(str) {
			return envelope.user.name + ": " + str;
		}))));
	}

	topic() {
		var envelope, strings;
		envelope = arguments[0], strings = 2 <= arguments.length ? slice.call(arguments, 1) : [];
		return this.bot.Room(envelope.room).topic(strings.join(" / "), (function(_this) {
			return function(err, data) {
				if (err != null) {
					return _this.robot.logger.error("Campfire topic error: " + err);
				}
			};
		})(this));
	}

	play() {
		var envelope, strings;
		envelope = arguments[0], strings = 2 <= arguments.length ? slice.call(arguments, 1) : [];
		return this.bot.Room(envelope.room).sound(strings.shift(), (function(_this) {
			return function(err, data) {
				if (err != null) {
					_this.robot.logger.error("Campfire sound error: " + err);
				}
				return _this.play.apply(_this, [envelope].concat(slice.call(strings)));
			};
		})(this));
	}

	locked() {
		var envelope, strings;
		envelope = arguments[0], strings = 2 <= arguments.length ? slice.call(arguments, 1) : [];
		if (envelope.message["private"]) {
			return this.send.apply(this, [envelope].concat(slice.call(strings)));
		} else {
			return this.bot.Room(envelope.room).lock((function(_this) {
				return function() {
					var args;
					args = 1 <= arguments.length ? slice.call(arguments, 0) : [];
					strings.push(function() {
						return setTimeout((function() {
							return _this.bot.Room(envelope.room).unlock();
						}), 3000);
					});
					return _this.send.apply(_this, [envelope].concat(slice.call(strings)));
				};
			})(this));
		}
	}

	run() {
		var bot, options, self, withAuthor;
		self = this;
		options = {
			token: process.env.HUBOT_CAMPFIRE_TOKEN,
			rooms: process.env.HUBOT_CAMPFIRE_ROOMS,
			account: process.env.HUBOT_CAMPFIRE_ACCOUNT
		};
		bot = new CampfireStreaming(options, this.robot);
		withAuthor = function(callback) {
			return function(id, created, room, user, body) {
				return bot.User(user, function(err, userData) {
					var author, userId;
					if (userData.user) {
						author = self.robot.brain.userForId(userData.user.id, userData.user);
						userId = userData.user.id;
						self.robot.brain.data.users[userId].name = userData.user.name;
						self.robot.brain.data.users[userId].email_address = userData.user.email_address;
						author.room = room;
						return callback(id, created, room, user, body, author);
					}
				});
			};
		};
		bot.on("TextMessage", withAuthor(function(id, created, room, user, body, author) {
			var message;
			if (bot.info.id !== author.id) {
				message = new TextMessage(author, body, id);
				message["private"] = bot["private"][room];
				return self.receive(message);
			}
		}));
		bot.on("EnterMessage", withAuthor(function(id, created, room, user, body, author) {
			if (bot.info.id !== author.id) {
				return self.receive(new EnterMessage(author, null, id));
			}
		}));
		bot.on("LeaveMessage", withAuthor(function(id, created, room, user, body, author) {
			if (bot.info.id !== author.id) {
				return self.receive(new LeaveMessage(author, null, id));
			}
		}));
		bot.on("TopicChangeMessage", withAuthor(function(id, created, room, user, body, author) {
			if (bot.info.id !== author.id) {
				return self.receive(new TopicMessage(author, body, id));
			}
		}));
		bot.on("LockMessage", withAuthor(function(id, created, room, user, body, author) {
			return bot["private"][room] = true;
		}));
		bot.on("UnlockMessage", withAuthor(function(id, created, room, user, body, author) {
			return bot["private"][room] = false;
		}));
		bot.Me(function(err, data) {
			var i, len, ref1, results, roomId;
			bot.info = data.user;
			bot.name = bot.info.name;
			ref1 = bot.rooms;
			results = [];
			for (i = 0, len = ref1.length; i < len; i++) {
				roomId = ref1[i];
				results.push((function(roomId) {
					return bot.Room(roomId).join(function(err, callback) {
						return bot.Room(roomId).listen();
					});
				})(roomId));
			}
			return results;
		});
		bot.on("reconnect", function(roomId) {
			return bot.Room(roomId).join(function(err, callback) {
				return bot.Room(roomId).listen();
			});
		});
		this.bot = bot;
		return self.emit("connected");
	}
}

module.exports = VkAdapter


src/streaming.js
const EventEmitter        = require('events').EventEmitter

class CampfireStreaming extends EventEmitter {

	constructor(options, robot1) {
		this.robot = robot1;
		if (!((options.token != null) && (options.rooms != null) && (options.account != null))) {
			this.robot.logger.error("Not enough parameters provided. I need a token, rooms and account");
			process.exit(1);
		}
		this.token = options.token;
		this.rooms = options.rooms.split(",");
		this.account = options.account;
		this.host = this.account + ".campfirenow.com";
		this.authorization = "Basic " + new Buffer(this.token + ":x").toString("base64");
		this["private"] = {};
	}

	Rooms (callback) {
		return this.get("/rooms", callback);
	}

	User (id, callback) {
		return this.get("/users/" + id, callback);
	}

	Me (callback) {
		return this.get("/users/me", callback);
	}

	Room(id) {
		var logger, self;
		self = this;
		logger = this.robot.logger;
		return {
			show: function(callback) {
				return self.get("/room/" + id, callback);
			},
			join: function(callback) {
				return self.post("/room/" + id + "/join", "", callback);
			},
			leave: function(callback) {
				return self.post("/room/" + id + "/leave", "", callback);
			},
			lock: function(callback) {
				return self.post("/room/" + id + "/lock", "", callback);
			},
			unlock: function(callback) {
				return self.post("/room/" + id + "/unlock", "", callback);
			},
			paste: function(text, callback) {
				return this.message(text, "PasteMessage", callback);
			},
			topic: function(text, callback) {
				var body;
				body = {
					room: {
						topic: text
					}
				};
				return self.put("/room/" + id, body, callback);
			},
			sound: function(text, callback) {
				return this.message(text, "SoundMessage", callback);
			},
			speak: function(text, callback) {
				var body;
				body = {
					message: {
						"body": text
					}
				};
				return self.post("/room/" + id + "/speak", body, callback);
			},
			message: function(text, type, callback) {
				var body;
				body = {
					message: {
						"body": text,
						"type": type
					}
				};
				return self.post("/room/" + id + "/speak", body, callback);
			},
			listen: function() {
				var headers, options, ref1, ref2, request;
				headers = {
					"Host": "streaming.campfirenow.com",
					"Authorization": self.authorization,
					"User-Agent": "Hubot/" + ((ref1 = this.robot) != null ? ref1.version : void 0) + " (" + ((ref2 = this.robot) != null ? ref2.name : void 0) + ")"
				};
				options = {
					"agent": false,
					"host": "streaming.campfirenow.com",
					"port": 443,
					"path": "/room/" + id + "/live.json",
					"method": "GET",
					"headers": headers
				};
				request = HTTPS.request(options, function(response) {
					var buf;
					response.setEncoding("utf8");
					buf = '';
					response.on("data", function(chunk) {
						var data, error, offset, part, results;
						if (chunk === ' ') {

						} else if (chunk.match(/^\s*Access Denied/)) {
							return logger.error("Campfire error on room " + id + ": " + chunk);
						} else {
							buf += chunk;
							results = [];
							while ((offset = buf.indexOf("\r")) > -1) {
								part = buf.substr(0, offset);
								buf = buf.substr(offset + 1);
								if (part) {
									try {
										data = JSON.parse(part);
										results.push(self.emit(data.type, data.id, data.created_at, data.room_id, data.user_id, data.body));
									} catch (error1) {
										error = error1;
										results.push(logger.error("Campfire data error: " + error + "\n" + error.stack));
									}
								} else {
									results.push(void 0);
								}
							}
							return results;
						}
					});
					response.on("end", function() {
						logger.error("Streaming connection closed for room " + id + ". :(");
						return setTimeout(function() {
							return self.emit("reconnect", id);
						}, 5000);
					});
					return response.on("error", function(err) {
						return logger.error("Campfire listen response error: " + err);
					});
				});
				request.on("error", function(err) {
					return logger.error("Campfire listen request error: " + err);
				});
				return request.end();
			}
		};
	}

	get(path, callback) {
		return this.request("GET", path, null, callback);
	}

	post(path, body, callback) {
		return this.request("POST", path, body, callback);
	}

	put(path, body, callback) {
		return this.request("PUT", path, body, callback);
	}

	request(method, path, body, callback) {
		var headers, logger, options, ref1, ref2, request;
		logger = this.robot.logger;
		headers = {
			"Authorization": this.authorization,
			"Host": this.host,
			"Content-Type": "application/json",
			"User-Agent": "Hubot/" + ((ref1 = this.robot) != null ? ref1.version : void 0) + " (" + ((ref2 = this.robot) != null ? ref2.name : void 0) + ")"
		};
		options = {
			"agent": false,
			"host": this.host,
			"port": 443,
			"path": path,
			"method": method,
			"headers": headers
		};
		if (method === "POST" || method === "PUT") {
			if (typeof body !== "string") {
				body = JSON.stringify(body);
			}
			body = new Buffer(body);
			options.headers["Content-Length"] = body.length;
		}
		request = HTTPS.request(options, function(response) {
			var data;
			data = "";
			response.on("data", function(chunk) {
				return data += chunk;
			});
			response.on("end", function() {
				var error;
				if (response.statusCode >= 400) {
					switch (response.statusCode) {
						case 401:
							throw new Error("Invalid access token provided");
							break;
						default:
							logger.error("Campfire HTTPS status code: " + response.statusCode);
							logger.error("Campfire HTTPS response data: " + data);
					}
				}
				if (callback) {
					try {
						return callback(null, JSON.parse(data));
					} catch (error1) {
						error = error1;
						return callback(null, data || {});
					}
				}
			});
			return response.on("error", function(err) {
				logger.error("Campfire HTTPS response error: " + err);
				return callback(err, {});
			});
		});
		if (method === "POST" || method === "PUT") {
			request.end(body, 'binary');
		} else {
			request.end();
		}
		return request.on("error", function(err) {
			return logger.error("Campfire request error: " + err);
		});
	}
}

module.exports = CampfireStreaming


.eslintrc.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# A VK.COM adapter for hubot.
#
# Copyright (C) 2016-2020 Volebo <dev@volebo.net>
# Copyright (C) 2016-2020 Maksim Koryukov <maxkoryukov@volebo.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

'use strict'


const path = require('path')

exports = module.exports = {
	'root': true,

	'env': {
		node: true
	},

	'parserOptions': {
		// 'parser': 'babel-eslint'
		'ecmaVersion': 2018,
	},

	'extends': [
		// 'plugin:vue/essential',
		// 'volebo',
		path.join(__dirname, 'node_modules', 'eslint-config-volebo', 'index.js'),
	],

	'overrides': [
		{
			'files': [
				'**/*.test.js',
				'**/*.spec.js'
			],

			'env': {
				mocha: true
			},

			'globals': {
				'filename2suitename': true,
				'tags': true,
			},

		}
	],
}
